### PR TITLE
Lock-free parts rename for ReplicatedMergeTree to avoid INSERT affect SELECT (resubmit)

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Lock-free parts rename to avoid INSERT affect SELECT (due to parts lock) (under normal circumstances with `fsync_part_directory`, QPS of SELECT with INSERT in parallel, increased 2x, under heavy load the effect is even bigger). Note, this only includes `ReplicatedMergeTree` for now

Under heavy load, or not so heavy but with fsync_part_directory=1,
rename could take longer then usual, and this will increase the time
that DataPartsLock will be held, while DataPartsLock is the bottleneck
for MergeTree (any SELECT requires it to obtain list of parts).

On one of production clusters I saw ~60 seconds with
fsync_part_directory=1.

So this patch moves rename out of critical section, by explicitly rename
the parts, to make it more clear when it is done without holding a lock.

Note, that this patch fixes the bottleneck only for ReplicatedMergeTree,
MergeTree is not that simple, and should be done separately.

#### Patched

```
┌─log_comment─────────────────────────────┬───QPS─┬─queries─┬────lock_wait_avg_us─┬─duration_q95─┐
│ lockfree-rename.90977.mt_without_fsync  │ 384.3 │    3843 │ 0.41217798594847777 │            2 │
│ lockfree-rename.90977.mt_with_fsync     │ 422.2 │    4222 │ 0.06395073424917101 │            2 │ # this part had been reverted!
│ lockfree-rename.90977.rmt_without_fsync │ 368.6 │    3686 │  1.9834508952794356 │            2 │
│ lockfree-rename.90977.rmt_with_fsync    │   417 │    4170 │ 0.22853717026378897 │            2 │
└─────────────────────────────────────────┴───────┴─────────┴─────────────────────┴──────────────┘
```

#### Upstream

```
┌─log_comment─────────────────────────────┬───QPS─┬─queries─┬───lock_wait_avg_us─┬─duration_q95─┐
│ lockfree-rename.91765.mt_with_fsync     │ 196.2 │    1962 │  2436.145259938838 │           19 │
│ lockfree-rename.91765.rmt_with_fsync    │ 203.9 │    2039 │ 2214.7631191760665 │           18 │
│ lockfree-rename.91765.rmt_without_fsync │ 394.6 │    3946 │  2.725798276735935 │            2 │
│ lockfree-rename.91765.mt_without_fsync  │ 411.2 │    4112 │  15.97932879377432 │            2 │
└─────────────────────────────────────────┴───────┴─────────┴────────────────────┴──────────────┘
```

_Original - https://gist.github.com/azat/7df0643cebb4ca9835d97734b18647b3_

Original PR: https://github.com/ClickHouse/ClickHouse/pull/61973 (cc @alexey-milovidov )
Revert: https://github.com/ClickHouse/ClickHouse/pull/64899 (cc @alesapin )

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 036dd3142ee94d2446d3c1d630c56f9c0e64724b with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64955/036dd3142ee94d2446d3c1d630c56f9c0e64724b/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by azat** [comment]
I've added benchmark into the description, which basically shows that without any other load (the ideal scenario) QPS of SELECTs that executed in parallel with INSERT, from table with `fsync_part_directory=1` decreased ~2x (q95 goes from 2ms to 19ms and waiting for the data parts lock increased to 2ms)

And this, is without any load, numbers will got higher under load.

**Comment by azat** [comment]
Test failures **does** related:
- `00276_sample`
- `00054_merge_tree_partitions`

File: src/Storages/MergeTree/MergeTreeSink.cpp [hunk 1]
```diff
@@ -186,7 +186,17 @@ void MergeTreeSink::finishDelayedChunk()
...
+            /// FIXME: renames for MergeTree should be done under the same lock
+            /// to avoid removing extra covered parts after merge.
+            ///
+            /// Image the following:
+            /// - T1: all_2_2_0 is in renameParts()
+            /// - T2: merge assigned for [all_1_1_0, all_3_3_0]
+            /// - T1: renameParts() finished, part had been added as Active
+            /// - T2: merge finished, covered parts removed, and it will include all_2_2_0!
+            ///
+            /// Hence, for now rename_in_transaction is false.
```
Comment (azat) [inline src/Storages/MergeTree/MergeTreeSink.cpp:198]: This was the problem, for now, I've reverted this lock-free renames for plain MergeTree

**Comment by azat** [comment]
Can someone take a look please?

**Comment by alexey-milovidov** [comment]
The error in `02479_mysql_connect_to_self` is haunting me - either we have too low timeouts for MySQL in tests and it is normal if sometimes the connection takes a long time, or the MySQL library does not handle EINTR from profiler, or something else. Let's take a look.

**Comment by alexey-milovidov** [comment]
I don't understand why Unit tests failed... we have to check.

**Comment by azat** [comment]
>I don't understand why Unit tests failed... we have to check.

a) it is a bug in gdb (images have gdb 12.1) - https://sourceware.org/bugzilla/show_bug.cgi?id=29185
b) it is some kind of issue of unit tests likely, maybe the spot instance got a ACPI shutdown signal? or they 

So I think they CI retry logic should be improved

**Comment by alexey-milovidov** [comment]
@azat, then make a pull request with update to a new gdb version.

**Comment by alexey-milovidov** [comment]
We need to fix this test: https://s3.amazonaws.com/clickhouse-test-reports/64955/70691ed3e9ff39d0b4f02e55319e3054e55ad933/stateless_tests__debug__[1_5].html

**Comment by azat** [comment]
>@azat, then make a pull request with update to a new gdb version.

I don't think that this will help, since by some reason it got SIGTERM, and my guess is due to ACPI shutdown, like I wrote, the CI retry logic should be improved I guess

**Comment by azat** [comment]
>We need to fix this test: [https://s3.amazonaws.com/clickhouse-test-reports/64955/70691ed3e9ff39d0b4f02e55319e3054e55ad933/stateless_tests__debug__[1_5].html](https://s3.amazonaws.com/clickhouse-test-reports/64955/70691ed3e9ff39d0b4f02e55319e3054e55ad933/stateless_tests__debug__%5B1_5%5D.html)

I will take a look, but it has nothing to do with this changes, it is better to track it here - https://github.com/ClickHouse/ClickHouse/issues/50911

**Comment by alexey-milovidov** [comment]
Let's resolve conflicts and merge (if it will be mergeable).

**Comment by azat** [comment]
OK, let's try to rebase to resolve some tests issues

**Comment by azat** [comment]
Tidy was broken, let's do one more rebase

**Comment by alexey-milovidov** [comment]
Just in case, I don't recommend rebasing, because it involves `git push --force` and then my conflict resolutions in the private repository are rewritten.

**Comment by azat** [comment]
>Just in case, I don't recommend rebasing, because it involves git push --force and then my conflict resolutions in the private repository are rewritten.

Got it (didn't knew about how conflict resolution works with private fork, now it is clear, thanks)

**Comment by azat** [comment]
@alexey-milovidov BTW likely I can fix this, if you will tell me for which commit you fixed the confclits

**Comment by azat** [comment]
Kind ping, can someone take a look please?

**Comment by azat** [comment]
>Performance Comparison (aarch64) [2/4] — Failed to parse the report.

- https://github.com/ClickHouse/ClickHouse/pull/67795

**Comment by azat** [comment]
Can someone please take a look?

**Comment by azat** [comment]
>Builds — 34/35 artifact groups are OK

- https://github.com/ClickHouse/ClickHouse/issues/68266
- https://github.com/ClickHouse/ClickHouse/pull/68267

**Comment by alexey-milovidov** [comment]
I don't know why no one reviewed this PR yet.
I started reading it... The code is fairly complex. Maybe I will finish reading it tomorrow.

**Comment by alexey-milovidov** [comment]
I already reviewed it a while ago :)

**Pull request merged**